File: src/ray.cpp
#include "ray.h"
#include "tuple.h"
#include "matrix.h"
#include "comparinator.h"

Ray::Ray()
{
	this->mbrOrigin = Point(0,0,0);
	this->mbrDirection = Vector(0,0,0);
}

Ray::Ray(Point origin, Vector direction)
{
	this->mbrOrigin = origin;
	//std::cout << this->origin.x << std::endl;
	this->mbrDirection = direction;
}
Point Ray::getPosition(double time)
{
	return mbrOrigin + mbrDirection * time;
}
bool Ray::checkEqual(Ray other)
{
	Comparinator varComp = Comparinator();
	return varComp.checkTuple(mbrOrigin, other.mbrOrigin) && varComp.checkTuple(mbrDirection, other.mbrDirection);
}
Ray Ray::transform(Matrix matrix)
{
	return Ray(matrix * this->mbrOrigin, matrix * mbrDirection);
}
Ray Ray::transform(TranslationMatrix matrix)
{
	return Ray(matrix * this->mbrOrigin, matrix * mbrDirection);
}
Ray Ray::transform(ScalingMatrix matrix)
{
	return Ray(matrix * this->mbrOrigin, matrix * mbrDirection);
}

File: src/Ray.cs
using LibTuple;
using LibMatrix;
using LibComparinator;
namespace LibRay;

public class Ray {
	public SpaceTuple _fieldOrigin;
	public SpaceTuple _fieldDirection;

	public Ray()
	{
		_fieldOrigin = new Point(0,0,0);
		_fieldDirection = new Vector(0,0,0);
	}

	public Ray(SpaceTuple origin, SpaceTuple direction)
	{
		_fieldOrigin = origin;
		_fieldDirection = direction;
	}
	public Ray(Ray argOther) {
		_fieldOrigin = argOther._fieldOrigin;
		_fieldDirection = argOther._fieldDirection;
	}
	public SpaceTuple GetPosition(double time)
	{
		SpaceTuple varDelta = _fieldDirection * time;
		SpaceTuple varRes = _fieldOrigin + varDelta;
		return varRes;
	}
	public bool CheckEqual(Ray other)
	{
		Comparinator varComp = new Comparinator();
		return varComp.CheckTuple(_fieldOrigin, other._fieldOrigin) && varComp.CheckTuple(_fieldDirection, other._fieldDirection);
	}
	public void ChangeTransform(Matrix matrix)
	{
		_fieldOrigin = matrix * _fieldOrigin;
		_fieldDirection = matrix * _fieldDirection;
		// return new Ray(matrix * _fieldOrigin, matrix * _fieldDirection);
	}
	public Ray GetTransform(Matrix matrix)
	{
		return new Ray(matrix * _fieldOrigin, matrix * _fieldDirection);
	}
	public void RenderConsole() {
		Console.WriteLine("Ray::renderConsole()");
		Console.WriteLine("Ray::renderConsole()::_fieldOrigin(");
		_fieldOrigin.RenderConsole();
		Console.WriteLine("Ray::renderConsole()::_fieldDirection(");
		_fieldDirection.RenderConsole();
	}
};